decred/decred/dcr/account.py

```diff
@@ -831,19 +1088,169 @@ def calcBalance(self, tipHeight=None):
+        if ticketTxid not in self.ticketDB:
+            raise Exception(
+                "spending tx came for ticket that was not found in the database: {}".format(
+                    ticketTxid
+                )
+            )
+        # sometimes this data cannot be fetched yet by dcrdata
+        tinfo = self.spentTicketInfo(ticketTxid)
+        if not tinfo:
+            if isVote:
+                tinfo = TicketInfo.fromVote(self.blockchain, tx)
+            else:
+                tinfo = TicketInfo.fromRevocation(self.blockchain, tx)
+
+        ticket = self.ticketDB[ticketTxid]
+        ticket.tinfo = tinfo
+        self.ticketDB[ticketTxid] = ticket
+
+        self.signals.spentTickets()
+
+    def updateSpentTickets(self):
+        """
+        Update voted and revoked tickets. ui.SPENT_TICKETS_SIGNAL is called if
+        there was an update.
+        """
+        # live tickets
+        mempoolTickets = self.mempoolTickets()
+        tickets = self.liveTickets()
+        tickets.update(mempoolTickets)
+        # live tickets as stored in the database
+        unspentTickets = self.unspentTickets()
+
+        # Only update if we have reason to believe ticket statuses have changed.
+        # TODO: also compare ticket statuses
+        if set(tickets.keys()) == set(unspentTickets.keys()):
+            return
+
+        log.info("ticket statuses have changed, updating spent tickets")
+        votedTickets = self.votedTickets()
+        revokedTickets = self.revokedTickets()
+        addrs = [pool.purchaseInfo.ticketAddress for pool in self.stakePools]
+        # All tickets. txs are populated with unknown spent tickets during sync
+        txids = list(mempoolTickets.keys())
+        for addr in addrs:
+            if addr in self.txs:
+                txids.extend(self.txs[addr])
+
+        for txid in set(txids):
+            # no need to check confirmed revocations and votes
+            if txid in votedTickets or txid in revokedTickets:
+                if self.ticketDB[txid].tinfo.status != "unconfirmed":
+                    continue
+            if txid in tickets:
+                # no need to check live tickets that we know about and have been mined
+                if txid in unspentTickets and unspentTickets[txid].isLiveTicket():
+                    continue
+                log.info("found unspent ticket: {}".format(txid))
+                ticket = tickets[txid]
+                # Estimate txs fee. All ticket purchases for pools are about 540
+                # bytes.
+                ticket.tinfo.purchaseTxFee = txscript.calcMinRequiredTxRelayFee(
+                    self.blockchain.relayFee(), 540
+                )
+                self.ticketDB[txid] = unspentTickets[txid] = ticket
+                continue
+
+            tx = self.blockchain.tx(txid)
+            # Vout 0 determines whether this is a ticket.
+            txout = tx.txOut[0]
+            if txscript.isStakeSubmissionScript(0, txout.pkScript):
```

Comment:
If this evaluates to `False`, doesn't that indicate an error has occured somewhere?

Reply:
some of the txids associated with the pool address are not tickets.  There are also votes and revocations.